proposals/dgl-operator-proposal.md

```diff
@@ -0,0 +1,248 @@
+
+[Native DGL distributed training](https://docs.dgl.ai/guide/distributed.html) is based on data parallelism, and we (360 AI Infra team) applied this fully distributed approach on operator/controller philosophy for Kubernetes. This proposal is aimed at defining what the DGL Operator should behave, and add it to Kubeflow.
+
+## Goals
+
+A Kubeflow user should be able to run training using DGL as easily as then can using Tensorflow or PyTorch. This proposal is centered around a Kubernetes operator for DGL. A user should be able to run both single worker job with DGL (1 launcher and 1 worker) and distributed training jobs with DGL (1 launcher and multiple workers).
+
+This proposal defines the following:
+
+- A DGL operator
+- A way to deploy the DGL Operator
+- A single worker DGL example
+- A distributed (multiple workers) DGL example
+
+## Non-Goals
+
+Currently, for the scope of this proposal, we won't be addressing the method for serving the model.
+
+## API
+
+### Custom Resource Definition
+
+The CRD yaml example is as following:
+
+```yaml
+apiVersion: qihoo.net/v1alpha1
+kind: DGLJob
+metadata:
+  name: dgl-graphsage
+  namespace: dgl-operator
+spec:
+  cleanPodPolicy: OnCompletion
+  dglReplicaSpecs:
+    Launcher:
+      replicas: 1
+      template:
+        spec:
+          containers:
+          - image: dgloperator/graphsage:v0.1.0
+            name: dgl-graphsage
+            command:
+            - dglrun
+            args:
+            - --graph-name
+            - graphsage
+            # partition arguments
+            - --partition-entry-point
+            - code/load_and_partition_graph.py
+            - --num-partitions
+            - "2"
+            # training arguments
+            - --train-entry-point
+            - code/train_dist.py
+            - --num-epochs
+            - "1"
+            - --batch-size
+            - "1000"
+    Worker:
+      replicas: 2
+      template:
+        spec:
+          containers:
+          - image: dgloperator/graphsage:v0.1.0
+            name: dgl-graphsage
+```
+
+- The `cleanPodPolicy` can be optionally configured as `Always` / `Never` / `OnFailure` / `OnCompletion`, indicating whether to delete the pod when the task is terminated. Preferred `Never` in debugging and `OnCompletion` in production.
+
+- The content of `Launcher` and `Worker` are followed the `PodTemplateSpec`. Users can be free to add more native key-values according to the spec.
```

Comment:
Looks like we can reuse the common spec in https://github.com/kubeflow/common

Reply:
Ongoing development is based on pure Kubebuilder v2, not engaged in Kubeflow common spec definition or job controller. But we are glad to reuse them.